File: 3-leetcode/01-easy/94-binary-tree-inorder-traversal.js
// Soltuion with DFS
const inorderTraversal = function (root) {
  let result = [];

  function DFS(node) {
    if (!node) return;
    DFS(node.left);
    result.push(node.val);
    DFS(node.right);
  }

  DFS(root);
  return result;
};

// Solution with Stack
const inorderTraversalStack = function (root) {
  let result = [];
  let stack = [];
  let curr = root;

  while (curr || stack.length) {
    while (curr) {
      stack.push(curr);
      curr = curr.left;
    }
    curr = stack.pop();
    result.push(curr.val);
    curr = curr.right;
  }
  return result;
};

File: 3-leetcode/01-easy/cpp/94-binary-tree-inorder-traversal.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<int> answer;
  void dfs(TreeNode *node)
  {
    if (!node)
      return;
    dfs(node->left);
    answer.push_back(node->val);
    dfs(node->right);
  }

  vector<int> inorderTraversal(TreeNode *root)
  {
    dfs(root);
    return answer;
  }

  vector<int> inorderTraversalIteratively(TreeNode *root)
  {
    vector<int> answer;
    stack<TreeNode *> stk;
    while (root || !stk.empty())
    {
      while (root)
      {
        stk.push(root);
        root = root->left;
      }
      root = stk.top();
      stk.pop();
      answer.push_back(root->val);
      root = root->right;
    }
    return answer;
  }
};